test/tests/i2c.c

```diff
@@ -0,0 +1,102 @@
+#include "FreeRTOS.h"
+#include "stm32xx_hal.h"
+#include "task.h"
+
+// Functions used by main
+void Clock_Config(void);
+static void MX_GPIO_Init(void);
+static void MX_I2C1_Init(void);
+void TestTask(void *argument);
+
+// Variables required for scheduling i2c_hal
+I2C_HandleTypeDef hi2c1;
+StaticTask_t task_buffer;
+StackType_t taskStack[configMINIMAL_STACK_SIZE];
+
+/**
+ * @brief  The application entry point.
+ * @retval int
+ */
+int main(void) {
+  HAL_Init();
+  MX_GPIO_Init();
+  MX_I2C1_Init();
+
+  // Create the task, passing in a priority level and stack size
+  // Then start the scheduler, which will begin executing tasks
+  xTaskCreateStatic(&TestTask, "Test", configMINIMAL_STACK_SIZE, NULL,
+                    tskIDLE_PRIORITY + 2, taskStack, &task_buffer);
+  vTaskStartScheduler();
+
+  // If the scheduler starts successfully, code below will not run
+  while (1) {
+  }
+}
+
+/**
+ * @brief I2C1 Initialization Function
+ * @param None
+ * @retval None
+ */
+static void MX_I2C1_Init(void) {
+  hi2c1.Instance = I2C1;
+  #ifdef STM32L4xx_HAL_SMBUS_H
+  #if PCLK1_FREQ == 16000000
+  hi2c1.Init.Timing = 0x10909CEC;
+  #elif PCLK1_FREQ == 32000000
+  hi2c1.Init.Timing = 0x20D09DE7;
+  #elif PCLK1_FREQ == 48000000
+  hi2c1.Init.Timing = 0x30F0A8E3;
+  #else
+    #error "Unsupported PCLK1_FREQ"
+  #endif
+  #endif
+  #ifdef STM32F4xx_HAL_SMBUS_H
+  hi2c1.Init.ClockSpeed = 100000;
+  hi2c1.Init.DutyCycle = I2C_DUTYCYCLE_2;
+  #endif
+  hi2c1.Init.OwnAddress1 = 0;
+  hi2c1.Init.AddressingMode = I2C_ADDRESSINGMODE_7BIT;
+  hi2c1.Init.DualAddressMode = I2C_DUALADDRESS_DISABLE;
+  hi2c1.Init.OwnAddress2 = 0;
+  hi2c1.Init.GeneralCallMode = I2C_GENERALCALL_DISABLE;
+  hi2c1.Init.NoStretchMode = I2C_NOSTRETCH_DISABLE;
+
+  __HAL_RCC_I2C1_CLK_ENABLE();
+  HAL_I2C_Init(&hi2c1);
+}
+
+/**
+ * @brief GPIO Initialization Function
+ * @param None
+ * @retval None
+ */
+static void MX_GPIO_Init(void) {
+  // GPIO Ports Clock Enable
+  __HAL_RCC_GPIOB_CLK_ENABLE();
+
+  GPIO_InitTypeDef GPIO_InitStruct = {0};
+
+  // Configure I2C SCL and SDA pins
+  GPIO_InitStruct.Pin =
+      GPIO_PIN_8 | GPIO_PIN_9;  // Example for I2C1 SCL and SDA pins on STM32F4
+  GPIO_InitStruct.Mode = GPIO_MODE_AF_OD;
+  GPIO_InitStruct.Pull = GPIO_NOPULL;
```

Comment:
Look at the other BSPs for the MspInit() function which is called by `HAL_{peripheral}_Init()`. The idea is that it handles basic GPIO initialization and powers up clocks. This way the BSP can handle this instead of user code.